blockio-api/src/System/FS/BlockIO/API.hs

```diff
@@ -214,6 +254,23 @@ hAdviseAll hbio h advice = hAdvise hbio h 0 0 advice -- len=0 implies until the
 hDropCacheAll :: HasBlockIO m h -> Handle h -> m ()
 hDropCacheAll hbio h = hAdviseAll hbio h AdviceDontNeed
 
+{-------------------------------------------------------------------------------
+  Storage synchronisation
+-------------------------------------------------------------------------------}
+
+synchroniseFile :: MonadThrow m => HasFS m h -> HasBlockIO m h -> FsPath -> m ()
+synchroniseFile hfs hbio path = do
+    b <- FS.doesFileExist hfs path
+    if b then
+      FS.withFile hfs path (FS.ReadWriteMode FS.AllowExisting) $ hSynchronise hbio
+    else
+      throwIO $ FS.ioToFsError (FS.mkFsErrorPath hfs (FS.mkFsPath [])) ioerr
```

Comment:
I presume the file existence test is to provide better exceptions? Since presumably `withFile` would throw an exception if the file does not exist.

If that's the reason, in principle the better approach is not to test for specific things (like the file not existing) but to just catch any io error and rethrow an improved error (e.g. with the error location set).

Reply:
The reason I'm explicitly throwing an error is subtle:

* `hOpen` and therefore `withFile` will throw errors when we use `ReadMode` and the file does not exist. So if we could use `ReadMode` then your suggestion could be applied, but...
* For `hSynchronise` on Windows, we have to open the file in (read-)write mode, and ...
* Using `hOpen` and therefore `withFile` in (read-)write mode requires passing an [`AllowExisting`](https://hackage.haskell.org/package/fs-api-0.3.0.1/docs/System-FS-API-Types.html#t:AllowExisting) argument. This determines how to handle existing or non-existing files, and it currently only has two options: `AllowExisting` and `MustBeNew`. The former will silently create the file if it does not exist, and the latter will throw an error if it *does* exist. So it currently doesn't allow throwing an error when the file does not exist

I agree with your suggestion to rethrow errors, but we'd have to modify `fs-api` first. I'll add a TODO explaining this

Reply:
Done in 8a07a26fe7eceb6f7797cee8d92ee3994918fba4